doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+terminal to invoke it.
+
+## Using the new template
+
+Now you should be able to use the new template by running (after compiling/installing Elektra again):
+
+```
+kdb gen example user userkeys
+```
+
+This should produce the file `userkeys.txt`, whose content should match the output of `kdb ls user`.
+
+## Advanced concepts
+
+Lastly, we will discuss a few more advanced concepts.
+
+### Switching delimiters
+
+Switching the delimiters is fully supported and already in use in the high-level API template to allow for easier formatting.
+
+### Parameters
+
+Sometimes you want to incorporate user input beyond the contents of the input keyset into the code-generation. For example, if you generate
+a web page, you may need to now where it is going to be hosted, so that you can generate correct links. Another example would be letting
+the user choose the name of a generated function, when generating C code.
+
+The first case would be required parameter. Without the domain name, we cannot generate the links. The second example meanwhile may be an
+optional parameter. We could just use a sensible default name for the function, since it doesn't matter much which function the user will
+call in their own code.
+
+Our framework supports both required and optional parameters. To define the parameters of your template use the `parameters` argument of the
+`GenTemplate` constructor. This parameter is a map from strings to bools. The keys are the parameter names and the values define, whether
+the parameter is required (`true` means required).
+
+Our two examples could use `{ { "domain_name", true } }` and `{ { "function_name", false } }` respectively.
+
+To access the parameter value call one of the `getParameter` overloads. One takes a default value, the other takes a map of values and
+verifies that one of the given values has been chosen. For more information see the relevant code documentation. There is also
+`getBoolParameter` which a specialised version for boolean parameters. It accepts only `0` and `1` as values.
+
+### Using partials
+
+The use of partials is bit more involved than in other mustache frameworks. All the partial files for template `X` must be placed in the
+folder `src/tools/kdb/gen/templates/X` and must use the file extension `.mustache`. Apart from the the filename can be chosen arbitrarily.
+
+To use the partial named `Y` (i.e. the file `src/tools/kdb/gen/templates/X/Y.mustache`) you must use this mustache command:
+
+```
+{{> partial.enum.c }}
+```
+
+The prefix `partial.` is required by the framework, if you omit it, there will be an error.
+
+### Custom escape functions
+
+By default mustache escapes values for use in HTML (unless `{{{ name }}}` or `{{& name }}` is used). Since most of our templates are not in
+fact HTML, the escape function can be customised. You simply have to override `GenTemplate::escapeFunction`. For an example see
+`ElektraGenTemplate::escapeFunction`, it is designed for C code instead of HTML.
```

Comment:
In which file?

Reply:
`elektragen.cpp`, I expected that its fully qualified name would be enough to find the function.

Reply:
Of course you know where it is but the reader might need to use `find`.